pkg/table-rule-selector/trie_selector.go

```diff
@@ -0,0 +1,461 @@
+
+	return errors.Trace(err)
+}
+
+func (t *trieSelector) insertSchema(schema string, rule interface{}, replace bool) error {
+	_, err := t.insert(t.root, schema, rule, replace)
+	if err != nil {
+		return errors.Annotate(err, "insert into schema selector")
+	}
+
+	return nil
+}
+
+func (t *trieSelector) InsertTable(schema, table string, rule interface{}, replace bool) error {
+	schemaEntity, err := t.insert(t.root, schema, nil, false)
+	if err != nil {
+		return errors.Annotate(err, "insert into schema selector")
+	}
+
+	if schemaEntity.nextLevel == nil {
+		schemaEntity.nextLevel = newNode()
+	}
+
+	_, err = t.insert(schemaEntity.nextLevel, table, rule, replace)
+	if err != nil {
+		return errors.Annotate(err, "insert into table selector")
+	}
+
+	return nil
+}
+
+// if rule is nil, just extract nodes
+func (t *trieSelector) insert(root *node, pattern string, rule interface{}, replace bool) (*item, error) {
+	var (
+		n           = root
+		hadAsterisk = false
+		entity      *item
+	)
+
+	for i := range pattern {
+		if hadAsterisk {
+			return nil, errors.Errorf("pattern %s is invaild", pattern)
+		}
+
+		switch pattern[i] {
+		case asterisk:
+			entity = n.asterisk
+			hadAsterisk = true
+		case question:
+			entity = n.question
+		default:
+			entity = n.characters[pattern[i]]
+		}
+		if entity == nil {
+			entity = &item{}
+			switch pattern[i] {
+			case asterisk:
+				n.asterisk = entity
+			case question:
+				n.question = entity
+			default:
+				n.characters[pattern[i]] = entity
+			}
+		}
+		if entity.child == nil {
+			entity.child = newNode()
+		}
+		n = entity.child
+	}
+
+	if rule != nil {
+		if !replace && entity.rule != nil {
+			return nil, errors.AlreadyExistsf("pattern %s", pattern)
+		}
+		entity.rule = rule
+		t.clearCache()
+	}
+
+	return entity, nil
+}
+
+// Match implements Selector's interface.
+func (t *trieSelector) Match(schema, table string) RuleSet {
+	// try to find schema/table in cache
+	t.RLock()
+	cacheKey := quoateSchemaTable(schema, table)
+	rules, ok := t.cache[cacheKey]
+	t.RUnlock()
+	if ok {
+		return rules.clone()
+	}
+
+	matchedSchemaResult := &macthedResult{
+		nodes: make([]*node, 0, 4),
+		rules: make(RuleSet, 0, 4),
+	}
+	rules = nil
+
+	// find matched rules
+	t.Lock()
+	defer t.Unlock()
+	t.matchNode(t.root, schema, matchedSchemaResult)
+
+	// not found matched rules in schema level
+	if matchedSchemaResult.empty() {
+		t.addToCache(cacheKey, nil)
+		return nil
+	}
+
+	rules = append(rules, matchedSchemaResult.rules...)
+
+	// only need to find schema level matched rule
+	if len(table) == 0 {
+		t.addToCache(cacheKey, rules)
+		return rules.clone()
+	}
+
+	for _, si := range matchedSchemaResult.nodes {
+		matchedTableResult := &macthedResult{
+			rules: make(RuleSet, 0, 4),
+		}
+		// find matched rules in table level
+		t.matchNode(si, table, matchedTableResult)
+		rules = append(rules, matchedTableResult.rules...)
+	}
+	// not found matched rule in table level, return mathed rule in schema level
```

Comment:
if not found matched rule in table level, it's only a partial match, do we return `rules` rather than `nil`?

Reply:
do u mean we should return nil? in our usage scenarios, the schema-level rules also have some usage significance, so these are left to the external user

Reply:
I am very confused here, because of the unification of concepts and interface, I have never wanted to break down schema and table to return

Reply:
seems just keep return `rules` is ok. fine